### ISSUE TITLE
reportUndefinedVariable diagnostic in named decorator parameters with semantic highlighting enabled

### ISSUE THREAD

**Issue by zgoda** [microsoft/pylance-release#322]
## Environment data

-   Language Server version: 2020.9.0
-   OS and version: Linux Ubuntu 20.04 x86_64
-   Python version (& distribution if applicable, e.g. Anaconda): 3.8.5, Anaconda

## Expected behaviour

`autouse` is recognised as valid `pytest.fixture()` argument

## Actual behaviour

"autouse" is not defined message in problems output

## Code Snippet / Additional information

```python
    @pytest.fixture(autouse=True)
    def _set_up(self, mocker):
        fake_connparams = {
            'username': 'test-user',
            'password': 'test-password',
            'host': 'test-host',
            'port': 1234
        }
```

![Zrzut ekranu z 2020-09-04 14-20-51](https://user-images.githubusercontent.com/141999/92239012-9804ca80-eeba-11ea-8b43-f5f3e9e808ec.png)

**Comment by zgoda** [comment]
Also, 2020.9.0 made the colours in One Dark Pro theme just crazy.
![Zrzut ekranu z 2020-09-04 14-29-25](https://user-images.githubusercontent.com/141999/92239324-12354f00-eebb-11ea-9427-a77444888331.png)

**Comment by zgoda** [comment]
#323 has more on this crazy code highlight, even with the same colour scheme.

**Comment by jakebailey** [comment]
We can talk about the highlighting on that issue (disable semantic highlighting if you don't like it), but the squiggle on the parameter looks to be a diagnostic bug.

**Comment by jakebailey** [comment]
Does the diagnostic go away if you disable semantic highlighting? E.g. disable it for all files or just Python:

```json
{
    "editor.semanticHighlighting.enabled": false
}
```

```json
{
    "[python]": {
        "editor.semanticHighlighting.enabled": false
    }
}
```

**Comment by zgoda** [comment]
Disabling semantic highlighting has no effect on this, although colours now look acceptable.

![Zrzut ekranu z 2020-09-04 19-25-26](https://user-images.githubusercontent.com/141999/92269821-3e65c580-eee5-11ea-9b3c-7aec84d6cf3a.png)
![Zrzut ekranu z 2020-09-04 19-25-52](https://user-images.githubusercontent.com/141999/92269825-3f96f280-eee5-11ea-85e0-926102366e26.png)
![Zrzut ekranu z 2020-09-04 19-29-12](https://user-images.githubusercontent.com/141999/92269828-3f96f280-eee5-11ea-9c4f-2ca7baf00a4b.png)

**Comment by erictraut** [comment]
Semantic highlighting should have no impact on diagnostics or vice versa.

I'm trying to replicate your first (simplest) example, and I'm not seeing the same behavior. Let's see if we can determine what's different between your confirmation and mine.

If you click on "fixture" and press F12 (go to definition), where does it take you? For me, it takes me to "site-packages/_pytest/fixtures.py", which looks like this:

![Screen Shot 2020-09-04 at 10 38 39 AM](https://user-images.githubusercontent.com/7040122/92270415-dee2c800-ee9a-11ea-8771-46fcddacc48a.png)

I'm trying to determine what declaration of `fixture` you're using because it appears that your declaration doesn't have a `autouse` parameter defined.

**Comment by zgoda** [comment]
~/.local/lib/python3.8/site-packages/_pytest/fixtures.py and this is wrong, this is conda env so this should be like ~/miniconda3/envs/envname/lib/python3.8/site-packages/_pytest/fixtures.py

**Comment by jakebailey** [comment]
The key thing here is that it says "undefined variable" which is a use-time diagnostic. Naming a parameter in a function call / decorator should not be triggering that diagnostic, as that name is not a use, regardless of if it doesn't actually exist in the function decl.

**Comment by zgoda** [comment]
This is somehow related to semantic highlighting setting. If I disable it then the issue disappears and reappears once I enable it. Works for both user and workspace settings.

**Comment by zgoda** [comment]
The tooltip on issue marker in the editor also shows wrong signature for `autouse: Any` (should be: `bool`).

**Comment by jakebailey** [comment]
This might be another bug in the analysis in the background thread that does the semantic tokens; we saw a similar issue during development that was fixed. That's why I bought it up.

RE: the `Any`, hovering over a named parameter looks at the function declaration for the type, not what you pass in, so it's going to the decorator, not finding `autouse`, and then saying it's `Any` (as in "I don't know").

**Comment by EonYang** [comment]
I got the same issue after updating to 2020.9.0.

## Screenshots:


<img width="984" alt="Screen Shot 2020-09-04 at 3 30 41 PM" src="https://user-images.githubusercontent.com/9332910/92284213-97c2fa00-eecf-11ea-8f8e-10a8a2c903e1.png">

⇡ *The error message*

<br />

<img width="1483" alt="Screen Shot 2020-09-04 at 3 31 05 PM" src="https://user-images.githubusercontent.com/9332910/92284319-bb864000-eecf-11ea-9e52-579ffc3ffacb.png">

⇡ *Note that Pylance is able to pull out the function signature which clearly includes the **undefined** argument*

<br />

<img width="945" alt="Screen Shot 2020-09-04 at 3 30 20 PM" src="https://user-images.githubusercontent.com/9332910/92286157-38b3b400-eed4-11ea-9f5a-dba9b72bf6c7.png">

⇡ *It's even able to give suggestions.*

## Downgrade to 2020.8.3 solved my issue.

<img width="1483" alt="Screen Shot 2020-09-04 at 4 53 30 PM" src="https://user-images.githubusercontent.com/9332910/92286210-61d44480-eed4-11ea-89f6-1713f22d214f.png">

**Comment by erictraut** [comment]
I found the cause of this problem. The new semantic token provider was calling the core type evaluator on a parse node that was not an expression. In particular, it was passing the node within a call expression that corresponds to the name of a named parameter — for example, `autouse` in the example above. I discovered that the hover provider was also contained the same bug, and there might be others as well. Rather than try to find and fix all of the cases where this is occurring, I made the type evaluator more robust so it detects this case and handles it appropriately.

Here is the commit within the pyright repo:
https://github.com/microsoft/pyright/commit/d538318a0152785d8585f0f58d74a7c8b02e8742

@judej, this bug is apparently affecting many Pylance users, so we might want to do an interim release of Pylance that cherry-picks this commit from pyright.
